chapter_generative-adversarial-networks/gan.md

```diff
@@ -42,65 +42,112 @@ $$min_D max_G \{ -E_{x \sim \text{Data}} log D(\mathbf x) - E_{z \sim \text{Nois
 
 Many of the GANs applications are in the context of images. As a demonstration purpose, we are going to content ourselves with fitting a much simpler distribution first. We will illustrate what happens if we use GANs to build the world's most inefficient estimator of parameters for a Gaussian. Let us get started.
 
-```{.python .input  n=1}
+```{.python .input}
 %matplotlib inline
 from d2l import mxnet as d2l
 from mxnet import autograd, gluon, init, np, npx
 from mxnet.gluon import nn
 npx.set_np()
 ```
 
+```{.python .input}
+#@tab pytorch
+%matplotlib inline
+from d2l import torch as d2l
+import torch
+from torch.utils.data import DataLoader
+from torch.autograd import Variable
+from torch import nn
+```
+
 ## Generate some "real" data
 
 Since this is going to be the world's lamest example, we simply generate data drawn from a Gaussian.
 
-```{.python .input  n=2}
+```{.python .input}
 X = np.random.normal(size=(1000, 2))
 A = np.array([[1, 2], [-0.1, 0.5]])
 b = np.array([1, 2])
 data = X.dot(A) + b
 ```
 
+```{.python .input}
+#@tab pytorch
+X = torch.normal(0.0, 1, (1000, 2))
+A = torch.tensor([[1, 2], [-0.1, 0.5]])
+b = torch.tensor([1, 2])
+data = torch.mm(X, A) + b
+```
+
 Let us see what we got. This should be a Gaussian shifted in some rather arbitrary way with mean $b$ and covariance matrix $A^TA$.
 
-```{.python .input  n=3}
+```{.python .input}
 d2l.set_figsize()
 d2l.plt.scatter(data[:100, 0].asnumpy(), data[:100, 1].asnumpy());
 print(f'The covariance matrix is\n{np.dot(A.T, A)}')
 ```
 
-```{.python .input  n=4}
+```{.python .input}
+#@tab pytorch
+d2l.set_figsize()
+d2l.plt.scatter(data[:100, 0].numpy(), data[:100, 1].numpy());
+print(f'The covariance matrix is\n{torch.mm(A.T, A)}')
+```
+
+```{.python .input}
 batch_size = 8
 data_iter = d2l.load_array((data,), batch_size)
 ```
 
+```{.python .input}
+#@tab pytorch
+batch_size = 8
+data_iter = DataLoader(data, batch_size=batch_size)
+```
+
 ## Generator
 
 Our generator network will be the simplest network possible - a single layer linear model. This is since we will be driving that linear network with a Gaussian data generator. Hence, it literally only needs to learn the parameters to fake things perfectly.
 
-```{.python .input  n=5}
+```{.python .input}
 net_G = nn.Sequential()
 net_G.add(nn.Dense(2))
 ```
 
+```{.python .input}
+#@tab pytorch
+net_G = nn.Sequential()
+net_G.add_module("Linear", nn.Linear(2, 2))
```

Comment:
```suggestion
net_G = nn.Sequential(nn.Linear(2, 2))
```
We are defining all the layers at once in Sequential for consistency.

Reply:
I think add_module is more simliar to add in mxnet.

Reply:
Both way is ok.https://discuss.pytorch.org/t/when-to-use-add-module-function/10534.